src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs

```diff
@@ -0,0 +1,205 @@
+            {
+                var fileInfo = fileProvider.GetFileInfo(subpath);
+                if (fileInfo != null && fileInfo.Exists)
+                {
+                    return fileInfo;
+                }
+            }
+            return new NotFoundFileInfo(subpath);
+        }
+
+        /// <summary>
+        /// Enumerate a directory at the given path, if any.
+        /// </summary>
+        /// <param name="subpath">The path that identifies the directory</param>
+        /// <returns>Contents of the directory. Caller must check Exists property.
+        /// The content is a merge of the contents of the provided <see cref="IFileProvider"/>.
+        /// When there is multiple <see cref="IFileInfo"/> with the same Name property, only the first one is included on the results.</returns>
+        public IDirectoryContents GetDirectoryContents(string subpath)
+        {
+            var directoryContents = new CombinedDirectoryContents(_fileProviders, subpath);
+            return directoryContents;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="IChangeToken"/> for the specified <paramref name="filter"/>.
+        /// </summary>
+        /// <remarks></remarks>
+        /// <param name="pattern">Filter string used to determine what files or folders to monitor. Example: **/*.cs, *.*, subFolder/**/*.cshtml.</param>
+        /// <returns>An <see cref="IChangeToken"/> that is notified when a file matching <paramref name="filter"/> is added, modified or deleted.
+        /// The change token will be notified when one of the change token returned by the provided <see cref="IFileProvider"/> will be notified.</returns>
+        public IChangeToken Watch(string pattern)
+        {
+            // Watch all file providers
+            var activeChangeTokens = new List<IChangeToken>();
+            foreach (var fileProvider in _fileProviders)
+            {
+                var changeToken = fileProvider.Watch(pattern);
+                if (changeToken != null && changeToken.ActiveChangeCallbacks)
+                {
+                    activeChangeTokens.Add(changeToken);
+                }
+            }
+
+            // There is no change token with active change callbacks
+            if (activeChangeTokens.Count == 0)
+            {
+                return NoopChangeToken.Singleton;
+            }
+            var combinedFileChangeToken = new CombinedFileChangeToken(activeChangeTokens);
+            return combinedFileChangeToken;
+        }
+
+        private class CombinedDirectoryContents : IDirectoryContents
+        {
+            private readonly Lazy<List<IDirectoryContents>> _directoriesContents;
+            private readonly Lazy<List<IFileInfo>> _files;
+            private readonly Lazy<bool> _exists;
+
+            public CombinedDirectoryContents(IFileProvider[] fileProviders, string subpath)
+            {
+                _directoriesContents =new Lazy<List<IDirectoryContents>>(() =>
```

Comment:
This is in response to @davidfowl feedback: https://github.com/aspnet/FileSystem/pull/142#discussion_r45708539